Initial implementation of pocli (#1)
veraison/pocli: README.md

`pocli` is the command line interface for Veraison policy management service.
Policies are used to provide deployment-specific rules for amending and/or
augmenting attestation results generated by an attestation scheme evidence

> handler. Please see Version services [policy documentation] for more details of

Reviewer: ```suggestion
handler. Please see Veraison services [policy documentation] for more details of
```

how policies are used in Veraison.

[policy documentation]: https://github.com/veraison/services/tree/main/policy#readme

## Policy management overview

A policy is always associated with an [attestation scheme]. Veraison can keep
track of multiple policies for a single scheme, however at most one of them
can be "active" at a time. The active policy is the one that is actually
utilised by Veraison verification service.

When adding a new policy, `pocli` will automatically activate it (unless told
otherwise via a flag), however it is also possible to activate previously added

> policies (e.g. to roll back a bad update).

Reviewer: ```suggestion
otherwise via a flag), however it is also possible to activate any previously added
policy (e.g. to roll back a bad update).
```

  specified on the command line using `-H`/`--host` option.
- **port**: the port on which the Veraison management service is listening.
  This can also be specified on the command line using `-p`/`--port` option.

See `misc/example-config.yaml` for an example configuration file.

## Usage examples

All policy management operations are performed with respect to a specific
attestation scheme. Thus most `pocli` commands take the scheme name as their
first argument. To see a list of valid scheme names for the service, you can
execute

    ./pocli schemes

A new policy can be created by passing the name of the attestation scheme to
which it is to be added and the path to the file containing the policy rules to
the `create` subcommand:

    ./pocli create PSA_IOT misc/example-PSA_IOT-policy.rego

This will also automatically activate the new policy (which can be suppressed
with `-d`/`--dont-activate` option). Optionally, the policy can be given a name
with `-n`/`--name`. If successful, `pocli` will output a JSON dump of the newly
created policy.

You can obtain the currently active policy using the `get` subcommand:

    ./pocli get PSA_IOT

> This will write a JSON dump of the currently active policy to stdout. If there

Reviewer: ```suggestion
This will write a JSON dump of the currently active policy to stdout. If there is
```

no active policy for the scheme, `polci` will output an error to that effect.
`-o`/`--output` option can be used to write the JSON to a file instead of
stdout. Alternatively, `-w`/`--write-rules` can be used to write the rules to
the file specified with this option. That rules file can then be used with
`create` subcommand to create a new version of a policy.

The `get` subcommand can also be used to obtain a specific policy associated
with a scheme, rather than the active one specifying its UUID:

    ./pocli get PSA_IOT 94659974-2580-11ee-8fab-0242c0a84005

You can use `list` subcommands to get a list of all policies that have been
created for a scheme:

    ./pocli list PSA_IOT

As with the `get` subcommand, `-o`/`--output` option may be used to write the
resulting JSON array to a file.

A previously created policy may be activated with `activate` subcommand:

>     ./pocli get PSA_IOT 94659974-2580-11ee-8fab-0242c0a84005

Reviewer: ```suggestion
    ./pocli activate PSA_IOT 94659974-2580-11ee-8fab-0242c0a84005
```

As only one policy per scheme may be active at a time, this will also
deactivate the previously active policy, if there is one.

You can also deactivate all policies associated with a scheme, so that no

> policy would be applied when the evidence for that scheme is evaluated by the

Reviewer: ```suggestion
policy would be applied when the evidence for that scheme is appraised by the
```

`pocli` is the command line interface for Veraison policy management service.
Policies are used to provide deployment-specific rules for amending and/or
augmenting attestation results generated by an attestation scheme evidence

> handler. Please see Veraison services [policy documentation] for more details of

Reviewer: ```suggestion
augmenting attestation results generated by an attestation scheme (for example
PSA or CCA) . Please see Veraison services [policy documentation] for more details of
```

- **ctime** is the creation timestamp.
- **name** is a descriptive label associated with the policy. This can be
  specified when creating a policy. If not specified, the default name
  `"default"` is used. Names do not need to be unique -- multiple policies can
  have the same name. This can be used to group policies that are different
  version of the same logical policy.
- **type** indicates the policy engine that will be used to evaluate the
  policy, and therefore the syntax of the policy rules. Currently, the only
  supported type is `"opa"`.
- **rules** are the policy rules that will be applied to the attestation
  result. The format of the rules depends on the policy type. See the Veraison
  services [OPA policy documentation] for details on how to write "opa"
  policy rules. You can also seen an example in
  `misc/example-PSA_IOT-policy.rego`.
- **active** is a boolean value indicating whether this policy is the current
  active policy for the scheme. At most one policy is active, however it is
  also possible for a scheme to have no active policies.

[OPA policy documentation]: https://github.com/veraison/services/blob/main/policy/README.opa.md

## Configuration

> Configuration for `pocli` is specified using YAML markup language. `polci`

Reviewer: ```suggestion
Configuration for `pocli` is specified using YAML markup language. `pocli`
```

`-c`/`--config` option.

The following configuration options are supported:

- **host**: the host name of the Veraison management service. This can also be
  specified on the command line using `-H`/`--host` option.
- **port**: the port on which the Veraison management service is listening.
  This can also be specified on the command line using `-p`/`--port` option.

See `misc/example-config.yaml` for an example configuration file.

## Usage examples

All policy management operations are performed with respect to a specific
attestation scheme. Thus most `pocli` commands take the scheme name as their
first argument. To see a list of valid scheme names for the service, you can
execute

    ./pocli schemes

A new policy can be created by passing the name of the attestation scheme to
which it is to be added and the path to the file containing the policy rules to
the `create` subcommand:

    ./pocli create PSA_IOT misc/example-PSA_IOT-policy.rego

This will also automatically activate the new policy (which can be suppressed
with `-d`/`--dont-activate` option). Optionally, the policy can be given a name
with `-n`/`--name`. If successful, `pocli` will output a JSON dump of the newly

> created policy.

Reviewer: is the JSON dump on the console or to a supplied file name as argument? kindly clarify

Author: There is no option to specify an output file as arguement so it can only be to stdout.

You can obtain the currently active policy using the `get` subcommand:

    ./pocli get PSA_IOT

This will write a JSON dump of the currently active policy to stdout. If there

> no active policy for the scheme, `polci` will output an error to that effect.

Reviewer: ```suggestion
no active policy for the scheme, `pocli` will output an error to that effect.
```

`-o`/`--output` option can be used to write the JSON to a file instead of
stdout. Alternatively, `-w`/`--write-rules` can be used to write the rules to
the file specified with this option. That rules file can then be used with
`create` subcommand to create a new version of a policy.

The `get` subcommand can also be used to obtain a specific policy associated
with a scheme, rather than the active one specifying its UUID:

    ./pocli get PSA_IOT 94659974-2580-11ee-8fab-0242c0a84005

> You can use `list` subcommands to get a list of all policies that have been

Reviewer: ```suggestion
You can use `list` subcommand to get a list of all policies that have been
```
